src/app/clusters/scenes-server/SceneHandlerImpl.cpp

```diff
@@ -16,10 +16,186 @@
  */
 
 #include <app/clusters/scenes-server/SceneHandlerImpl.h>
+#include <app/util/endpoint-config-api.h>
+#include <app/util/odd-sized-integers.h>
 
 namespace chip {
 namespace scenes {
 
+namespace {
+
+template <int ByteSize, bool IsSigned>
+using OddSizedInteger        = app::OddSizedInteger<ByteSize, IsSigned>;
+using ConcreteAttributePath  = app::ConcreteAttributePath;
+using AttributeValuePairType = app::Clusters::ScenesManagement::Structs::AttributeValuePair::Type;
+
+/// ConvertDefaultValueToWorkingValue
+/// @brief Helper function to convert a byte array to a value of the given type.
+/// @param EmberAfDefaultAttributeValue & defaultValue
+/// @return Value converted to the given working type
+template <typename Type>
+Type ConvertDefaultValueToWorkingValue(const EmberAfDefaultAttributeValue & defaultValue)
+{
+    if (sizeof(Type) <= 2)
+    {
+        return static_cast<Type>(defaultValue.defaultValue);
+    }
+
+    Type sValue = 0;
+    memcpy(&sValue, defaultValue.ptrToDefaultValue, sizeof(Type));
+    return app::NumericAttributeTraits<Type>::StorageToWorking(sValue);
+}
+
+/// CapAttributeID
+/// Cap the attribute value based on the attribute's min and max if they are defined,
+/// or based on the attribute's size if they are not.
+/// @param[in] aVPair   AttributeValuePairType
+/// @param[in] metadata  EmberAfAttributeMetadata
+///
+template <typename Type>
+void CapAttributeID(AttributeValuePairType & aVPair, const EmberAfAttributeMetadata * metadata)
+{
+    using IntType     = app::NumericAttributeTraits<Type, CHIP_CONFIG_BIG_ENDIAN_TARGET>;
+    using WorkingType = typename IntType::WorkingType;
+
+    WorkingType maxValue;
+
+    if (metadata->IsBoolean())
+    {
+        aVPair.attributeValue = aVPair.attributeValue ? 1 : 0;
+        return;
+    }
+
+    // Check if the attribute type is signed
+    if (metadata->IsSignedIntegerAttribute())
+    {
+        maxValue = static_cast<WorkingType>((1ULL << (emberAfAttributeSize(metadata) * 8 - 1)) - 1);
+    }
+    else
+    {
+        maxValue = static_cast<WorkingType>((1ULL << (emberAfAttributeSize(metadata) * 8)) - 1);
+    }
+
+    // Check metadata for min and max values
+    if (metadata->HasMinMax())
+    {
+        const EmberAfAttributeMinMaxValue * minMaxValue = metadata->defaultValue.ptrToMinMaxValue;
+        WorkingType minVal                              = ConvertDefaultValueToWorkingValue<WorkingType>(minMaxValue->minValue);
+        WorkingType maxVal                              = ConvertDefaultValueToWorkingValue<WorkingType>(minMaxValue->maxValue);
+
+        // Cap based on minimum value
+        if (minVal > static_cast<WorkingType>(aVPair.attributeValue))
+        {
+            aVPair.attributeValue = static_cast<std::make_unsigned_t<WorkingType>>(minVal);
+            // We assume the max is >= min therefore we can return
+            return;
+        }
+
+        // Adjust maxValue if greater than the meta data's max value
```

Comment:
Only for the signed ones right? Since unsigned type min is always zero

Reply:
Either way, for now we treat overflowing the attribute type as greater than max since we cannot know the position of the sign bit (Nor the actual signedness that was received).